src/main/flight/altitude.c

```diff
@@ -219,61 +191,45 @@ void calculateEstimatedAltitude(timeUs_t currentTimeUs)
     previousTimeUs = currentTimeUs;
 
     static float vel = 0.0f;
-    static float accAlt = 0.0f;
-
     int32_t baroAlt = 0;
+    static int32_t baroAltOffset = 0;
+    static int32_t gpsAltOffset = 0;
+
+    int32_t gpsAlt = 0;
+    float gpsTrust = 0.3; //conservative default
+    bool haveBaroAlt = false;
+    bool haveGPSAlt = false;
 #ifdef USE_BARO
     if (sensors(SENSOR_BARO)) {
         if (!isBaroCalibrationComplete()) {
             performBaroCalibrationCycle();
             vel = 0;
-            accAlt = 0;
         } else {
             baroAlt = baroCalculateAltitude();
-            estimatedAltitude = baroAlt;
+            haveBaroAlt = true;
         }
     }
 #endif
 
-#ifdef USE_RANGEFINDER
-    if (sensors(SENSOR_RANGEFINDER) && rangefinderProcess(getCosTiltAngle())) {
-        int32_t rangefinderAlt = rangefinderGetLatestAltitude();
-        if (rangefinderAlt > 0 && rangefinderAlt >= rangefinderCfAltCm && rangefinderAlt <= rangefinderMaxAltWithTiltCm) {
-            // RANGEFINDER in range, so use complementary filter
-            float rangefinderTransition = (float)(rangefinderMaxAltWithTiltCm - rangefinderAlt) / (rangefinderMaxAltWithTiltCm - rangefinderCfAltCm);
-            rangefinderAlt = (float)rangefinderAlt * rangefinderTransition + baroAlt * (1.0f - rangefinderTransition);
-            estimatedAltitude = rangefinderAlt;
-        }
+#ifdef USE_GPS
+    if (sensors(SENSOR_GPS)) {
+        gpsAlt = gpsSol.llh.alt;
+        haveGPSAlt = true;
```

Comment:
Why don't you check if there's a GPS fix? Until there is a valid 3D fix `gpsSol.llh.alt` is meaningless.

Reply:
Good catch, will do.